### DISCUSSION TITLE
How To Build And Flash A Working Firmware Version

### DISCUSSION THREAD

**Discussion by felix-gohla** [solokeys/solo2#92] [Q&A]
Hey all,

tl;dr: can compile firmware, but self-built version seems to be broken

As part of a university project, we are evaluating the usage of FIDO2 in electronic locking systems. For that, we chose to buy some Solo v2, in order to extend the firmware a little (especially for CTAP 2.1 `largeBlob` extension).

We cloned the repo and after some fiddling around (reading the getting started HackMD and using the version of June / July, v1.0.8), we got it to compile. The cloned version as of now doesn’t seem to compile on a vanilla system due to some dependencies not being pinned (`ctap-types` and some others iirc).
We also tried fixing the current state. In the end, both variants compiled without a problem (with `make build-dev`).
Build system is a freshly installed Debian bullseye VM.

After that, we tried to flash the firmware (on macOS). Using the solo2-cli, we got the Soll into Bootloader mode and using `lpc55-host` we flashed the firmware (didn’t use mboot, because we were in macOS) So until this step, it everything worked ok. In that process, we probably did override all the flash.

But after rebooting the Solo, the LEDs were not lit, but the key was recognised by the solo2 cli, at least we could get it back into Bootloader mode. The NDEF app didn’t work, we couldn’t scan the NFC tag.

So we have some questions and would be really happy to receive some help from the Pros. ;)

1. By flashing the FW, did we override some important data? If yes, what  do we need to restore? We have a dump of the memory before flashing. I think the dev certs are gone now, as FIDO returns the default AAGUID as specified in in the `fido-Authenticator` repo. Something else?
2. How do we get the LEDs to light up? This seems to indicate, that we did something wrong. Did we choose the wrong board type during build?
3. Do you have some information on what exactly is preset in the hacker variant’s LPC55 settings (secure boot, user & factory settings, etc)? Or should we consult the data sheet for finding out ourselves? :)
4. What should we never do / touch in order to not brick the board?

Sorry for this being so long, but maybe we can help improving the documentation for the hackers to build their own FW. 🚀

**Comment by nickray** [comment]
Hey, thanks for the extensive feedback! Re. 4, the main issue is if your build somehow passes the bootloader, crashes early, and you can't get into bootloader again to flash a fix. I'd strongly recommend using NXP's dev kit for this reason. Will try to get back to the rest. We'd like full CTAP2.1 support, even though that exploded quite a bit... Also happy to chat to learn a bit more about your context (ping @nickray:solokeys.com on Matrix, or via email nicolas@solokeys.com if you'd like to set something up early Jan).
